### DISCUSSION TITLE
Multipart uploading is adding headers to file content

### DISCUSSION THREAD

**Discussion by dreinon** [encode/httpx#2004] [Potential Issue] [answered]
When I try to send a post request with any file with any mime-type, I get headers in the content of the file. Trying to debug the problem, I decided to use [curlify2](https://github.com/marcuxyz/curlify2) to turn the httpx request into a curl request, and calling the following:

```py
import httpx
import curlify2

with open("example.txt", "w") as f:
    f.write("This is an example")

client = httpx.Client()
content_type = "text/plain;charset=UTF-8"
r = client.build_request(
    method="POST",
    url="http://localhost:8081/just-an-example-url",
    files={"file": ("filename", open("example.txt", "rb"), content_type)},
    headers={"cacheControl": "3600", "content-type": content_type},
)

print(curlify2.to_curl(r))
```
prints the following output:
```bash
curl -X POST -H "host: localhost:8081" -H "accept: */*" -H "accept-encoding: gzip, deflate" -H "connection: keep-alive" -H "user-agent: python-httpx/0.21.1" -H "cachecontrol: 3600" -H "content-type: text/plain;charset=UTF-8" -H "content-length: 202" -d '--f6d68f1fb1980bf5a50d9e4c55ff6634
Content-Disposition: form-data; name="file"; filename="filename"
Content-Type: text/plain;charset=UTF-8

This is an example
--f6d68f1fb1980bf5a50d9e4c55ff6634--
' http://localhost:8081/just-an-example-url
```

We can see that the `-d` parameter, which is the body of the request, has the following content:
```
--f6d68f1fb1980bf5a50d9e4c55ff6634
Content-Disposition: form-data; name="file"; filename="filename"
Content-Type: text/plain;charset=UTF-8

This is an example
--f6d68f1fb1980bf5a50d9e4c55ff6634--
```
which is exactly the content I see in the sent file, whereas what I would expect it to be is:
```
This is an example
```

**Comment by dreinon** [comment]
@j0 @leynier @anand2312

**Comment by dreinon** [comment]
Hi @tomchristie, sorry to mention you. We are using `httpx` in the Supabase python client and we are experiencing this issue which lead us to errors in our storage client. Our client is widely used (not as much as `httpx` but there's quite a few people depending on it 🙂) , so we temporarily patched this feature in particular using `requests` but would appreciate a lot if you guys could take a look at it whenever you have a bit of time. Thanks!

**Reply by lovelydinosaur** [reply]
Heya - thanks for re-raising this - looks like it got lost in the mix. Putting it onto my TODO list to take a look at.

**Comment by lovelydinosaur** [comment, marked as answer]
I've not had a proper look into this, but quick thing to point out...

You're specifying the `Content-Type` of the overall request, with...

```python
headers={"cacheControl": "3600", "content-type": content_type},
```

Rather than just using that content type for the file upload.

The overall content type for the upload request ought to be `multipart/form-data` - which `httpx` will automatically populate for you. It's possible that this is what's causing the issue.

If it is, then we might have a look at handling this conflict more gracefully.

**Reply by dreinon** [reply]
Thanks for the quick answer, I'll take a look and get back to you with new 🙂

**Reply by dreinon** [reply]
That was indeed the issue we were facing. Thank you, much appreciated!

**Reply by lovelydinosaur** [reply]
Fab. Might be worth us looking into if we can guard against this better.

**Reply by dreinon** [reply]
Yes, might be a good idea to document this, since requests is okay with setting both headers, and thus, it's an api change between both libs.
